src/plugins/wpa/DataModel/ETW/QuicEtwEvent.cs

```diff
@@ -43,28 +46,79 @@ internal enum QuicEtwEventOpcode : byte
         Datapath
     }
 
-    internal class QuicEtwEvent : QuicEvent
+    internal unsafe ref struct EtwDataReader
     {
-        public override Guid Provider { get; }
+        private ReadOnlySpan<byte> Data;
 
-        public override QuicEventId ID { get; }
+        private readonly int PointerSize;
 
-        public override int PointerSize { get; }
+        internal EtwDataReader(void* pointer, int length, int pointerSize)
+        {
+            Data = new ReadOnlySpan<byte>(pointer, length);
+            PointerSize = pointerSize;
+        }
+
+        internal unsafe T ReadValue<T>() where T : unmanaged
+        {
+            T val = MemoryMarshal.Cast<byte, T>(Data)[0];
+            Data = Data.Slice(sizeof(T));
+            return val;
+        }
 
-        public override uint ProcessId { get; }
+        internal byte ReadByte() => ReadValue<byte>();
 
-        public override uint ThreadId { get; }
+        internal ushort ReadUShort() => ReadValue<ushort>();
 
-        public override ushort Processor { get; }
+        internal uint ReadUInt() => ReadValue<uint>();
 
-        public override Timestamp TimeStamp { get; }
+        internal ulong ReadULong() => ReadValue<ulong>();
 
-        public override QuicObjectType ObjectType { get; }
+        internal ulong ReadPointer()
+        {
+            return PointerSize == 8 ? ReadValue<ulong>() : ReadValue<uint>();
+        }
+
+        internal string ReadString()
+        {
+            var chars = new List<char>();
+            while (true)
+            {
+                byte c = ReadValue<byte>();
```

Comment:
Are these strings guaranteed to be ASCII only? If not, might be better to go into a List<byte>, then do Encoding.UTF8.GetString() to convert the UTF8 into UTF16. 

Also, I'd probably try catch ArgumentException here and return an empty string or "Invalid String" here if something goes wrong with this input. Try Catches are free in C# (Well the try part assuming nothing is caught is), and string parsing is definitely sensitive. But less important then the conversion part above.

Reply:
Actually don't worry about the try catch part. If we're going to do any of that, we should just do that in TryCreate, and return null if we catch, so we'll catch all parsing errors.

Reply:
All these guaranteed to be ASCII strings then?

Reply:
Yes.
```
            <data
                inType="win:AnsiString"
                name="ErrStr"
                />
```